### PR TITLE
Fix `qml.matrix` with multiple broadcasted operations in a tape

### PR DESCRIPTION
**Pull request by dwierichs**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
`qml.matrix` already was updated previously to work with broadcasted matrices, but the scenario of multiple
broadcasted operations or unbroadcasted operations _after_ a broadcasted one in, say, a tape was not considered. This scenario occurs in particular when decomposing
a broadcasted operation (which is supported for those operations that can be broadcasted).

**Description of the Change:**
Changes `qml.matrix` to manually run over the broadcasting dimension if both, a constituent of the tape and the accumulated matrix, are broadcasted.
For the previously uncovered cases, there are no changes except for a tiny slowdown due to checking of the logical branch (negligible). A dedicated Boolean was introduced to keep the overhead via dimension checking to a minimum.

**Benefits:**
Tapes/QNodes with multiple broadcasted operations and unbroadcasted operations after broadcasted ones are supported by `qml.matrix`.

**Possible Drawbacks:**
N/A 

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: doc/releases/changelog-dev.md
Comment (antalszava) [inline doc/releases/changelog-dev.md:156]: > as they are present in decompositions of broadcasted operations.

This reads a bit confusing, what does this refer to?

File: pennylane/ops/functions/matrix.py
Comment (antalszava) [inline pennylane/ops/functions/matrix.py:141]: Minor: suggest renaming to differentiate bwn. unitary of next op and the unitary whose matrix this method returns.
```suggestion
    broadcasted_result = False
```

File: pennylane/ops/functions/matrix.py
Comment (antalszava) [inline pennylane/ops/functions/matrix.py:144]: Minor: suggest renaming, the name `broadcasted_U` faintly tells me that the object is a broadcasted unitary whereas it's a boolean describing whether or not `U` was broadcasted.
```suggestion
        U_is_broadcasted = qml.math.ndim(U) == 3
```

File: pennylane/ops/functions/matrix.py
Comment (antalszava) [inline pennylane/ops/functions/matrix.py:158]: The previously suggested renamings (or better versions of those) should also help improve the readability of these lines. Right now, these names and the statements here can cause confusion.

File: pennylane/ops/functions/matrix.py
Comment (antalszava) [inline pennylane/ops/functions/matrix.py:148]: Could this be expressed using a `tensordot`? It seems like we're iterating over an axis and contracting another.

File: pennylane/ops/functions/matrix.py
Comment (antalszava) [inline pennylane/ops/functions/matrix.py:157]: Likely okay to set this once only.
```suggestion
        if not broadcasted_unitary and broadcasted_U:
```

File: tests/ops/functions/test_matrix.py
Comment (antalszava) [inline tests/ops/functions/test_matrix.py:221]: What if the number of angles differs for operations in the tape? Is there a validation check for that?

File: tests/ops/functions/test_matrix.py
Comment (antalszava) [inline tests/ops/functions/test_matrix.py:202]: Do we have a test for the case when the very first operation is broadcasted and there are non-broadcasted operations following?

Likely a variant to this and likely rearranging the order of ops here would yield that case already.

**Review by antalszava — CHANGES_REQUESTED** [review]
Overall looking great, requesting changes due to some of the naming here and there to improve readability. The rest are questions for my understanding.

File: doc/releases/changelog-dev.md
Comment (dwierichs) [reply doc/releases/changelog-dev.md:156]: Ah, agreed :D It refers to this problem/application popping up when we do
```python
with qml.tape.QuantumTape() as tape:
    qml.IsingXX(np.array([0.4, 2.1, -0.4]), wires=[1, 2]).decomposition()
qml.matrix(tape)
```
i.e. when we decompose a broadcasted operation and then want to compute the matrix of the resulting tape.

File: pennylane/ops/functions/matrix.py
Comment (dwierichs) [reply pennylane/ops/functions/matrix.py:141]: Should we rename `unitary_matrix` then as well? :)

File: pennylane/ops/functions/matrix.py
Comment (dwierichs) [reply pennylane/ops/functions/matrix.py:148]: I tried and failed :D 
The perspectives I found are:
1. We want to perform a standard matrix `dot`, which works with extra dimensions out of the box (e.g. `(3, 4, 4) @ (3, 4, 4)->(3,4,4)` exactly how we want), but it fails in cases where the broadcasting dimension is the same as the Hilbert space dimension (e.g. `(4, 4, 4) @ (4, 4, 4) -> (4, 4, 4, 4)` which is not what we want).
2. `einsum("...ij,..jk->...ik", A, B)` does exactly what we want, but it's terribly slow.
3. The approach I chose here. We kind of want to "multiply the matrices along the latter pair of axes while taking the diagonal w.r.t to the leading dimensions of both objects", if that makes any sense.

I'll check again whether there is magic in the `axes` kwarg of `tensordot` that I don't know yet.

File: tests/ops/functions/test_matrix.py
Comment (dwierichs) [reply tests/ops/functions/test_matrix.py:221]: Yes, tapes do not allow this :)

File: tests/ops/functions/test_matrix.py
Comment (dwierichs) [reply tests/ops/functions/test_matrix.py:202]: Good idea, will include! :)

File: doc/releases/changelog-dev.md
Comment (dwierichs) [reply doc/releases/changelog-dev.md]: Tried to make it more explicit/clear

File: pennylane/ops/functions/matrix.py
Comment (dwierichs) [reply pennylane/ops/functions/matrix.py]: I went for `result` instead of `unitary_matrix` and for `result_is_broadcasted`, in analogy to the below `U_is_broadcasted`.

File: doc/releases/changelog-dev.md
Comment (antalszava) [inline doc/releases/changelog-dev.md:157]:
```suggestion
  broadcasted operations and/or the decompositions of broadcasted operations.
```

File: doc/releases/changelog-dev.md
Comment (antalszava) [reply doc/releases/changelog-dev.md:157]: Thanks! Left a suggestion.

File: pennylane/ops/functions/matrix.py
Comment (antalszava) [reply pennylane/ops/functions/matrix.py:141]: Good question. That renaming is out of scope here. The clean way would be to rename that in a separate PR. Having said that, if we do want to make the change, I think we could do that here (provided it's a minor scope change, which it seems to be).

File: pennylane/ops/functions/matrix.py
Comment (antalszava) [reply pennylane/ops/functions/matrix.py:148]: Makes sense! No worries if it doesn't come around, maybe some of these findings could even be included as a comment (e.g., that `einsum` would be slower).

File: pennylane/ops/functions/matrix.py
Comment (dwierichs) [reply pennylane/ops/functions/matrix.py]: Time-wise the difference (as well as the overall consumed time) is really negligible, so we're down to readability. Any preference?

Comment (antalszava) [reply pennylane/ops/functions/matrix.py:157]: Hmm, I'd imagine the suggested version to be faster just because Python would evaluate the expression lazily (i.e., once `not broadcasted_unitary is False` `and broadcasted_U` is not being evaluated).

Readabilitywise, imho using improved names would make this block readable too.

Comment (dwierichs) [reply pennylane/ops/functions/matrix.py]: I agree, but because the loop only contains the assignment of a Boolean, it is similarly cheap as checking a Boolean, I think. In any case, this is about nanoseconds...

I updated the names, feel free to choose whatever is best here and I'll include it :)

File: pennylane/ops/functions/matrix.py
Comment (dwierichs) [reply pennylane/ops/functions/matrix.py]: Sorry, my comments were batched up in a review, hence the weird order. I think the scope is very small here, and it doesn't make sense to open a PR for this, imho.

File: doc/releases/changelog-dev.md
Comment (dwierichs) [reply doc/releases/changelog-dev.md]: I had already rewritten it a bit, to be more explicit in what I meant. Does this work now?

File: pennylane/ops/functions/matrix.py
Comment (dwierichs) [reply pennylane/ops/functions/matrix.py]: Aha! I just found out that `np.matmul` and `np.dot` have different broadcasting rules. `matmul` should be what we want :)

File: pennylane/ops/functions/matrix.py
Comment (dwierichs) [reply pennylane/ops/functions/matrix.py]: This is outdated now :)

File: doc/releases/changelog-dev.md
Comment (antalszava) [inline doc/releases/changelog-dev.md:161]:
```suggestion
  operations: the decomposition in general will contain one or multiple broadcasted
```

File: doc/releases/changelog-dev.md
Comment (antalszava) [inline doc/releases/changelog-dev.md:160]: Nice! :) This reads very well.

File: pennylane/ops/functions/matrix.py
Comment (antalszava) [inline pennylane/ops/functions/matrix.py:145]: :tada: :tada: :tada:

File: tests/ops/functions/test_matrix.py
Comment (antalszava) [inline tests/ops/functions/test_matrix.py:218]: Thanks for adding this!

**Review by antalszava — APPROVED** [review]
Looks good! :ok_hand: :clap:

Codecov seems to fail incorrectly. Once this is ready to be merged, let me know and the check can be overridden.

File: pennylane/ops/functions/matrix.py
Comment (dwierichs) [reply pennylane/ops/functions/matrix.py:145]: For future reference: An alternative here would be to make a `qml.math.matmul` modification analogous to `qml.math.tensordot`, that always calls coerce before calling the autoray `matmul` dispatcher. However, I thought this is overkill here.

**Comment by dwierichs** [comment]
Timing results seem to be shaky overall, but no systematic slowdown detected in the following.

Using this timing function:
```python
import time

def _timeit(fn, *args, n=100, include_first_run=False, timer=time.process_time):
    if include_first_run:
        start = timer()
        fn(*args)
        for _ in range(n):
            fn(*args)
        end = timer()
    else:
        fn(*args)
        start = timer()
        for _ in range(n):
            fn(*args)
        end = timer()
    return (end - start) / n
```
and the following generic tape:
```python
import scipy as sp
import pennylane as qml
from pennylane import numpy as pnp

def make_tape(num_wires):
    wires = list(range(num_wires))
    U = sp.stats.unitary_group.rvs(2**num_wires)
    with qml.tape.QuantumTape() as tape:
        [qml.RX(pnp.random.random(), w) for w in wires]
        [qml.CNOT(wires=[w, (w+1)%num_wires]) for w in wires]
        qml.QubitUnitary(U, wires=wires)
    return tape

times = []
for num_wires in tqdm(range(2, 11), total=9):
    reps = 10000//(2**num_wires)
    tape = make_tape(num_wires)
    times.append(_timeit(qml.matrix, tape, n=reps))
```
I obtained the following times:
```python
PR_times = [0.0005685069068000076, 0.0010907068408000213, 0.0014089999744000125, 0.002020689391025635, 0.0064590462948716666, 0.01647367947435896, 0.10729456238461574, 0.9513819041578939, 7.10383089222222]
master_times = [0.0006406608355999999, 0.0010427461496, 0.0014408085184000015, 0.0019606009166666673, 0.006181715461538457, 0.02278675079487181, 0.12348922397435899, 0.9145773105263159, 7.044265654999999]
```
or visually:
image

File: tests/ops/functions/test_matrix.py
Comment (mlxd) [inline tests/ops/functions/test_matrix.py:212]: I think it is fine for this instance (overall runtime is around 1ms), but explicitly creating the matrix here rather than kronning operations would be much more efficient. Even precomputing `np.kron(I, CNOT)` and `np.kron(S, H)` would reduce the number of ops needed here and locally cuts the time in half for me.

In this case , I'd suggest

```suggestion
        I_CNOT = np.kron(I, CNOT)
        S_H = np.kron(S, H)
        expected_matrix = [
            I_CNOT @ np.kron(I, S_H),
            -1j * I_CNOT @ np.kron(X, S_H),
            I_CNOT @ np.kron(RX(0.7), S_H),
        ]
```

File: tests/ops/functions/test_matrix.py
Comment (mlxd) [inline tests/ops/functions/test_matrix.py:232]:
```suggestion
        I_CNOT = np.kron(I, CNOT)
        S_H = np.kron(S, H)
        expected_matrix = [
            I_CNOT @ np.kron(I, S_H),
            -1j * I_CNOT @ np.kron(X, S_H),
            I_CNOT @ np.kron(RX(0.7), S_H),
        ]
```

File: tests/ops/functions/test_matrix.py
Comment (mlxd) [inline tests/ops/functions/test_matrix.py:255]: Same again here. I got 3x speedup by pre-allocating the repeated values. Also, `np.kron(I, np.kron(I, I))` can be more easily written by `qml.math.eye(8)`, avoiding the need to use kron product.

File: tests/ops/functions/test_matrix.py
Comment (mlxd) [inline tests/ops/functions/test_matrix.py:278]: And again here.

**Review by mlxd — CHANGES_REQUESTED** [review]
Thanks @dwierichs 
No major blockers from me, but I do have suggestions on imrpvoing the matrix building in the tests. Happy to approve once they are updated.

File: tests/ops/functions/test_matrix.py
Comment (dwierichs) [reply tests/ops/functions/test_matrix.py:212]: Oh, I did not look into optimization of this at all :) 
I'll implement these pre-computations for the file, so the kronecker products are only computed once overall.

**Review by mlxd — APPROVED** [review]
Great job @dwierichs 
Nothing further from me!

**Comment by dwierichs** [comment]
@antalszava This is ready to merge, may I ask you to override Codecov? :)

**Pull request merged**
